### PR TITLE
ENH: don't rebuild unchanged files

### PR DESCRIPTION
**Pull request by juliantaylor**
Only rebuild files that are newer than their corresponding object or if
older than their dependencies.

### CODE REVIEW

**Comment by juliantaylor** [comment]
I could have sworn distutils already did that and it stopped working at some point, am I remembering wrong?

**Comment by rgommers** [comment]
It has been doing that semi-reliably for a long time. Are you seeing complete rebuilds? What Python and setuptools versions?

**Comment by juliantaylor** [comment]
both python 2.7.13 and 3.5.3

**Comment by rgommers** [comment]
Hmm, I'll check tonight with my currently installed versions.

File: numpy/distutils/ccompiler.py [hunk 1]
```diff
@@ -19,6 +20,52 @@
...
+
+    Returns
+    -------
+    newer modification time of input
+    """
+    if not files:
+        return 0
+    newest = 0
+    for f in files:
+        newest = max(os.stat(f)[ST_MTIME], newest)
```
Comment (eric-wieser) [inline numpy/distutils/ccompiler.py]: Or just `os.stat(f).st_mtime`, and skip the `stat` import. Same below

Comment (eric-wieser) [inline numpy/distutils/ccompiler.py]: In fact, `return max(files, key=lambda f: os.stat(f).st_mtime)` could replace the last four lines

**Comment by rgommers** [comment]
``distutils`` does check file timestamps, you should see that if you rebuild the exact same commit twice in a row, then the second build does nothing (takes 1 sec). The issue (or feature?) you're likely seeing is that all files are rebuilt that together make up a single ``.so`` even if only one file was touched. This is due to the ``newer_group`` check [here](https://github.com/python/cpython/blob/master/Lib/distutils/ccompiler.py#L471) (I think - just reading the code, haven't verified) which is implemented [here](https://github.com/python/cpython/blob/master/Lib/distutils/dep_util.py#L54)

**Comment by juliantaylor** [comment]
distutils does not rebuild if nothing changed, but it rebuilds everything if one file changed. I thought it didn't do that in the past, but maybe I am imaging that as every other buildsystem works like that ...
This fixes this, but it has the drawback that the depends need to be set correctly as there is no automatic determination.

File: numpy/distutils/ccompiler.py (hunk 1 above)
Comment (juliantaylor) [inline numpy/distutils/ccompiler.py]: this returns the newest filename I guess the function should be called newest_file_time

Comment (eric-wieser) [inline numpy/distutils/ccompiler.py]: In which case, `return max(os.stat(f).st_mtime for f in files)` is better. Yep, rename sounds sensible

**Comment by juliantaylor** [comment]
should we add automatic dependency determination? its not difficult on unix platforms

**Comment by juliantaylor** [comment]
This is how it would look like.
If we are paranoid we could add a check for compiler support. I guess `CCompiler_customize` would be the right place for that?

**Comment by rgommers** [comment]
> distutils does not rebuild if nothing changed, but it rebuilds everything if one file changed. I thought it didn't do that in the past, but maybe I am imagining that as every other buildsystem works like that ...

If with "everything" you mean all files for a given extension module, then we are saying the same thing.

You can verify that different modules are independent

    time python setup.py build_ext -i   # ~1min 30s
    time python setup.py build_ext -i  # ~1s
    touch numpy/core/src/npysort/binsearch.c
    time python setup.py build_ext -i  # ~18s

This PR still looks fairly straightforward with the automatic dependency checking, so +1 to give it a go.

**Comment by rgommers** [comment]
How did you test this?

File: numpy/distutils/unixccompiler.py [hunk 2]
```diff
@@ -44,8 +44,9 @@ def UnixCCompiler__compile(self, obj, src, ext, cc_args, extra_postargs, pp_opts
             self.linker_so = llink_s.split() + opt.split()
 
     display = '%s: %s' % (os.path.basename(self.compiler_so[0]), src)
+    deps = ['-MMD', '-MF', obj + '.d']
```
Comment (rgommers) [inline numpy/distutils/unixccompiler.py]: this deserves a comment explaining what's going on here. I could figure it out, but those are not flags one uses every day

**Comment by rgommers** [comment]
> If we are paranoid we could add a check for compiler support. 

You mean checking that the rebuild is using the same compiler as the previous build?

**Comment by juliantaylor** [comment]
no, I meant checking if the compiler supports the commandline flags

**Comment by rgommers** [comment]
Ah, that's probably a healthy idea. We have people building on AIX, Raspberry Pi, and other assorted exotic platforms. ``CCompiler_customize`` should be the right place.

**Comment by juliantaylor** [comment]
Updated with more comments and a flag support check. It is skipped for gcc and clang which always support it as the customize is run on every extension so it actually costs measurable time in a single file recompile.

**Comment by juliantaylor** [comment]
probably release note worthy if we still want it in the next release

**Comment by rgommers** [comment]
CI is failing 

> probably release note worthy if we still want it in the next release

agreed

**Comment by juliantaylor** [comment]
ci looks good, it failed earlier due to msvc in distutils being stupid. Fixed that and I removed even trying on non posix.

**Comment by rgommers** [comment]
Okay, let's give this a try. In it goes. Thanks Julian!

**Comment by rgommers** [comment]
Want to send a follow up PR for release notes edit?

**Comment by charris** [comment]
This broke SciPy testing, see https://github.com/scipy/scipy/issues/7321. I don't know if this is a problem with the SciPy setup, but in any case i'm inclined to revert it until we can figure out what is going on.

File: numpy/distutils/ccompiler.py [hunk 3]
```diff
@@ -191,7 +230,8 @@ def CCompiler_compile(self, sources, output_dir=None, macros=None,
 
     def single_compile(args):
         obj, (src, ext) = args
-        self._compile(obj, src, ext, cc_args, extra_postargs, pp_opts)
+        if _needs_build(obj):
+            self._compile(obj, src, ext, cc_args, extra_postargs, pp_opts)
```
Comment (pv) [inline numpy/distutils/ccompiler.py]: The issue vs. #9035 is probably that `_needs_build` should also check if any compilation flags differ from the previous compilation.

Comment (charris) [inline numpy/distutils/ccompiler.py]: Could well be. I need to delete the SciPy build directory in order to run tests that show the problem if it was working in the previous build.

Comment (pv) [inline numpy/distutils/ccompiler.py]: Deleting might not help, because the same file is built two times with different options during the build.

Comment (charris) [inline numpy/distutils/ccompiler.py]: Just referring to my local build. Another problem I ran into was that in order to cythonize the source I needed to do a Python2 build first. I have cython installed for Python 3 but it goes under the name cython3.

Comment (juliantaylor) [inline numpy/distutils/ccompiler.py]: hm I though I tested scipy with this, apparently I forgot to run the tests

checking the arguments is simple enough, would be nicer to add the extension name to the dependency filename instead but it is probably clunky to get that from the compiler object.

**Pull request merged**